### PR TITLE
Device USB

### PR DESCRIPTION
**Pull request by tomponline**
Migration of usb device into the new device interface.

Part of #5819

Still WIP.

Includes https://github.com/lxc/lxd/pull/6050

### CODE REVIEW

File: lxd/device/usb.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+// validateConfig checks the supplied config for correctness.
+func (d *usb) validateConfig() error {
+	if d.instance.Type() != instance.TypeContainer {
+		return ErrUnsupportedDevType
+	}
+
+	rules := map[string]func(string) error{
+		"vendorid":  shared.IsAny,
```
Comment (stgraber) [inline lxd/device/usb.go]: We should add a validator for that and shared it with GPU too.
It's effectively 4 digit hex lowercase (unless we do case insensitive matches in which case we can allow both cases).

File: lxd/device/device_utils_usb.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
+package device
+
+import (
+	"fmt"
+	"path/filepath"
+	"strconv"
+)
+
+// devPath is the path where USB devices can be enumerated.
+const devPath = "/sys/bus/usb/devices"
```
Comment (stgraber) [inline lxd/device/device_utils_usb.go]: Should be usbDevPath

File: lxd/device/device_utils_usb.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+// USBDevice represents the properties of a USB device.
+type USBDevice struct {
```
Comment (stgraber) [inline lxd/device/device_utils_usb.go]: I wonder if we should consider pushing that data into the resources API, tracking USB devices in there directly as those fields look very similar to what we do for GPU, disks, ...

Probably not worth doing right now but if we see some demand for users trying to figure out USB devices on their systems, then moving to resources API would be the way to go.

**Comment by stgraber** [comment]
So I think there is a memory issue we need to look into and fix here.

You're effectively having the uevent package keep a reference to the event handling function for every single container and device which uses a USB device.

This function as it comes from a device struct keeps that struct active which in turn keeps the original instance struct active. That instance struct is a containerLXC struct which is then tied to a go-lxc container struct which holds a bunch of file descriptors.

This effectively means that for every single USB device we track, we end up with a full copy of all those structs and a bunch of open files.


To avoid this, we would need to have the event handling function be separate from any of those structs such that we don't keep any of the structs in memory and don't end up with a high memory usage + bunch of open files when facing containers using usb devices.

File: lxd/container_lxc.go [hunk 4]
```diff
@@ -1906,17 +1909,17 @@ func (c *containerLXC) deviceStart(deviceName string, rawConfig map[string]strin
 		return nil, fmt.Errorf("Device cannot be started when container is running")
 	}
 
-	runConfig, err := d.Start()
+	runConf, err := d.Start()
```
Comment (stgraber) [inline lxd/container_lxc.go]: Can you split the renaming of RunConfig to runConf into its own commit, would make review easier.

**Comment by tomponline** [comment]
@stgraber thanks for the feedback, in order to address the concern around memory consumption with holding the containerLXC and device structs in memory for USB event callbacks, I've reworked the implement as discussed on IRC.

I've not yet addressed your other points on validation, however if you can take a look at the new approach and check you're happy with it.

In summary the changes to approach are as follows:

- `uevents` package removed and USB event registration merged into `device` package - this is due to its dependency on `device` package already prevent calling it from inside `device` package causing circular references. There is still going to be a `uevents` package in the future but this will be focused on the actual netlink socket processing events rather than specific to USB device events.
- USB device now registers and unregisters its USB event hook directly rather than returning it via RunConfig, which was only done so it could get the container's context (no longer needed).
- The function it registers with the USB event handler is now self-contained and does not reference `d` or `d.instance`. 
- `Uevents` slice has been added to `RunConfig` in order to give a way for devices to instruct LXD to inject uevents to an instance.
- The `InstanceIdentifier` interface and `container` interface have had a `DeviceEventHandler(*device.RunConfig)` function added to them, this is used by the `USBRunHandlers` function after running the USB callback functions and being returned a non-empty RunConfig, to action the requests from the callback function inside the instance (after it loads the instance from the database and creates a fresh instantiation).

The last part is the key part, it adds a way for non-main packages to instantiate an instance and request that a `RunConfig` instruction set is actioned inside the instance.

File: lxd/device/usb.go (hunk 1 above)
Comment (tomponline) [reply lxd/device/usb.go]: Added a validator to usb and gpu.

**Pull request merged**